### PR TITLE
add uuid feature

### PR DESCRIPTION
**Pull request by Roms1383**
Add `uuid` feature flag, based on [uuid](https://github.com/uuid-rs/uuid) crate.

I originally opened, wrongly, a [feature request #524](https://github.com/uuid-rs/uuid/issues/524) there but then realized it should be part of this crate instead.

I tried to implement it by mimicking how the other supported data are implemented,

but I would like additional review because some inner mechanisms I didn't really grasp (at least it seemed, when I had the test passes something doesn't seem right, does it ?)

### CODE REVIEW

File: fake/src/faker/mod.rs [hunk 1]
```diff
@@ -226,3 +235,11 @@ pub mod automotive {
     }
 }
 
+pub mod uuid {
+    def_fakers! {
```
Comment (cksac) [inline fake/src/faker/mod.rs:239]: seems no impl for V1, V2, V3, V4 in fake/src/faker/impls/uuid.rs?
def_faker macro will generate config struct, then you implement Dummy for these structs.

File: fake/src/faker/mod.rs [hunk 2]
```diff
@@ -226,3 +235,11 @@ pub mod automotive {
     }
 }
 
+pub mod uuid {
```
Comment (cksac) [inline fake/src/faker/mod.rs:238]: missing #[cfg(feature = "uuid")]

File: fake/src/faker/mod.rs [hunk 3]
```diff
@@ -139,6 +147,7 @@ pub mod internet {
         MACAddress();
         Color();
         UserAgent();
+        Uuid(version: crate::faker::internet::UuidConfig);
```
[the remark this answers is not in the document]
Comment (Roms1383) [reply fake/src/faker/mod.rs:150]: I moved it to `fake/src/impls/uuid`, is this what you meant ?

**Comment by Roms1383** [comment]
@cksac I also went for a different approach, on a separate [derived branch](https://github.com/Roms1383/fake-rs/tree/feat/uuid-alt).
Especially as you mentioned about impl `dummy` and `dummy_with_rng` separately,
please give me some feedback :)

**Comment by cksac** [comment]
in general, we need to have following, which each item in def_faker will become an struct, so we don't need to have `UuidConfig` struct

fake/src/faker/mod.rs
```rust
#[cfg(feature = "uuid")]
pub mod uuid {
    def_faker! {
        UuidV1(uuid::v1::Timestamp, &'static [u8])
        UuidV3(uuid::Uuid, &'static [u8])
        UuidV4()
        UuidV5(uuid::Uuid, &'static [u8])
        UuidSeed(&'static [u8])
    }
}
```

fake/src/faker/impls/mod.rs
```rust
#[cfg(feature = "uuid")]
pub mod uuid;
```

fake/src/faker/impls/uuid.rs
```rust
impl<L: Data> Dummy<UuidV1> for uuid::Uuid {
    ...
}

impl<L: Data> Dummy<UuidV1> for String {
    ...
}

impl<L: Data> Dummy<UuidV3> for uuid::Uuid {
    ...
}

impl<L: Data> Dummy<UuidV4> for uuid::Uuid {
    ...
}

impl<L: Data> Dummy<UuidV5> for uuid::Uuid {
    ...
}

impl<L: Data> Dummy<UuidSeed> for uuid::Uuid {
    ...
}

impl<L: Data> Dummy<Faker> for uuid::Uuid {
    ...
}
```

**Comment by Roms1383** [comment]
okay I think it should be better now, what do you think ?

I'm still confused about a bunch of stuff :
- when trying to generate UUID v5 from slice, Rng is not implemented from `&[u8; 128]`
- not important but : why have both `faker/impls` and `impls` ? (just for my own sake of understanding 😅)

**Comment by Roms1383** [comment]
well actually I found out that I can do :
```rust
impl<L: Data> Dummy<UuidSeed<L>> for Uuid {
    fn dummy_with_rng<R: rand::Rng + ?Sized>(_: &UuidSeed<L>, rng: &mut R) -> Self {
        Self::from_u128(rng.gen())
    }

    fn dummy(config: &UuidSeed<L>) -> Self {
        // not sure about endianness here
        Self::from_u128(u128::from_be_bytes(config.1.try_into().expect("slice with incorrect length")))
    }
}
```
but not sure, so waiting for your feedback before committing

File: fake/src/faker/impls/uuid.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        todo!()
+        // let slice: &[u8; 128] = rng.gen();
+        // Self::from_slice(slice).expect("generate uuid::Uuid::from_slice")
+    }
+}
+
+#[allow(unused_variables)]
+impl Dummy<Faker> for Uuid {
+    fn dummy_with_rng<R: rand::Rng + ?Sized>(config: &Faker, rng: &mut R) -> Self {
+        todo!()
```
Comment (cksac) [inline fake/src/faker/impls/uuid.rs:120]: use `Uuid::from_u128(rng.gen())`?

File: fake/src/faker/impls/uuid.rs [hunk 5]
```diff
@@ -0,0 +1,126 @@
+use uuid::{Builder, Uuid, Variant, Version};
+
+use crate::faker::uuid::raw::*;
+use crate::locales::Data;
+use crate::{Dummy, Faker};
+
+impl<L: Data> Dummy<UuidV1<L>> for Uuid {
+    fn dummy_with_rng<R: rand::Rng + ?Sized>(config: &UuidV1<L>, rng: &mut R) -> Self {
```
Comment (cksac) [inline fake/src/faker/impls/uuid.rs:8]: UuidV1(timestamp, node_id), node_id is not used. Should we generated timestamp and node_id using rng and change UuidV1 to unit struct?

File: fake/src/faker/impls/uuid.rs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        let (ticks, _) = config.1.to_rfc4122();
+        let time_low = (ticks & 0xFFFF_FFFF) as u32;
+        let time_mid = ((ticks >> 32) & 0xFFFF) as u16;
+        let time_high_and_version = (((ticks >> 48) & 0x0FFF) as u16) | (1 << 12);
+        Uuid::from_fields(time_low, time_mid, time_high_and_version, &bytes)
+            .expect("generate uuid::Uuid::from_fields")
+    }
+
+    fn dummy(UuidV1(_, timestamp, node_id): &UuidV1<L>) -> Self {
+        Self::new_v1(*timestamp, node_id).expect("generate uuid::Uuid::new_v1")
```
Comment (cksac) [inline fake/src/faker/impls/uuid.rs:19]: seems this will always produce same Uuid?

File: fake/src/faker/impls/uuid.rs (hunk 4 above)
Comment (Roms1383) [reply fake/src/faker/impls/uuid.rs:120]: well, test passes indeed
should `dummy` for `Dummy<Faker>` also call `dummy_with_rng` ?

File: fake/src/faker/impls/uuid.rs (hunk 5 above)
Comment (Roms1383) [reply fake/src/faker/impls/uuid.rs:8]: Do you have something like this in mind ? :
(test passes also)
```rust
impl<L: Data> Dummy<UuidV1<L>> for Uuid {
    fn dummy_with_rng<R: rand::Rng + ?Sized>(config: &UuidV1<L>, rng: &mut R) -> Self {
        let UuidV1(_, _, node_id) = config;
        let ticks: u64 = rng.gen();
        let counter: u16 = rng.gen();
        let time_low = (ticks & 0xFFFF_FFFF) as u32;
        let time_mid = ((ticks >> 32) & 0xFFFF) as u16;
        let time_high_and_version = (((ticks >> 48) & 0x0FFF) as u16) | (1 << 12);
        let mut d4 = [0; 8];

        {
            d4[0] = (((counter & 0x3F00) >> 8) as u8) | 0x80;
            d4[1] = (counter & 0xFF) as u8;
        }

        d4[2..].copy_from_slice(&node_id);
        Uuid::from_fields(time_low, time_mid, time_high_and_version, &d4)
            .expect("generate uuid::Uuid::from_fields")
    }

    // ...
}
```
> (...) change UuidV1 to unit struct?
Not sure I understand this part

File: fake/src/faker/impls/uuid.rs (hunk 6 above)
Comment (Roms1383) [reply fake/src/faker/impls/uuid.rs:19]: yes indeed I'll check this out

**Comment by cksac** [comment]
It seems we don't need extra input for v1, v3, v4, v5? Something like below?

fake/src/faker/mod.rs
```rust
#[cfg(feature = "uuid")]
pub mod uuid {
    def_faker! {
        UuidV1();
        UuidV3();
        UuidV4();
        UuidV5();
    }
}
```

fake/src/faker/impls/uuid.rs
```rust
impl<L: Data> Dummy<UuidV1<L>> for Uuid {
    fn dummy_with_rng<R: rand::Rng + ?Sized>(_: &UuidV1<L>, rng: &mut R) -> Self {
        let ticks = Faker.fake_with_rng();
        let counter = Faker.fake_with_rng();
        let ts = uuid::v1::Timestamp::from_rfc4122(ticks, counter);
        let node_id [u8; 6] = Faker.fake_with_rng();
        Uuid::new_v1(ts, &node_id)
    }
}

impl<L: Data> Dummy<UuidV3<L>> for Uuid {
    fn dummy_with_rng<R: rand::Rng + ?Sized>(_: &UuidV3<L>, rng: &mut R) -> Self {
        Builder::from_bytes(rng.gen())
            .set_variant(Variant::RFC4122)
            .set_version(Version::Md5)
            .build()
    }
}


impl<L: Data> Dummy<UuidV4<L>> for Uuid {
    fn dummy_with_rng<R: rand::Rng + ?Sized>(_: &UuidV4<L>, rng: &mut R) -> Self {
        Builder::from_bytes(rng.gen())
            .set_variant(Variant::RFC4122)
            .set_version(Version::Random)
            .build()
    }
}

impl<L: Data> Dummy<UuidV5<L>> for Uuid {
    fn dummy_with_rng<R: rand::Rng + ?Sized>(_: &UuidV5<L>, rng: &mut R) -> Self {
        Builder::from_bytes(rng.gen())
            .set_variant(Variant::RFC4122)
            .set_version(Version::Sha1)
            .build()
    }
}
```

**Comment by Roms1383** [comment]
I've made the changes you suggested, ran the tests and repeatedly ran example fakers (different results obtained)

File: fake/examples/fakers.rs [hunk 7]
```diff
@@ -419,6 +419,23 @@ fn barecode_faker() {
     println!("{}", val);
 }
 
+fn uuid_faker() {
```
Comment (cksac) [inline fake/examples/fakers.rs:422]: missing #[cfg(feature = "uuid")]

Comment (Roms1383) [reply fake/examples/fakers.rs:423]: oh indeed 😅  thanks

**Comment by Roms1383** [comment]
> > ```
> > * not important but : why have both `faker/impls` and `impls` ? (just for my own sake of understanding 😅)
> > ```
> 
> `impls` for type primitive+std types and types don't require Locale Data
> `faker/impls` for type required Locale

Do the different uuids need Locale ?

**Comment by cksac** [comment]
I think uuid is not Locale specific

File: fake/src/lib.rs [hunk 8]
```diff
@@ -205,6 +205,9 @@ pub use impls::std::path::PathFaker;
 pub use impls::std::result::ResultFaker;
 pub use impls::std::string::StringFaker;
 
+#[cfg(feature = "uuid")]
+pub use impls::uuid;
+
```
Comment (Roms1383) [inline fake/src/lib.rs:210]: this one was required to be able to expose `uuid` from `impls` (private module)

**Review by Roms1383 — COMMENTED** [review]
if you don't mind, when everything is gonna be ok I would like to add some doc too

File: fake/src/lib.rs (hunk 8 above)
Comment (Roms1383) [reply fake/src/lib.rs:210]: @cksac so is this ok ?

**Comment by cksac** [comment]
> if you don't mind, when everything is gonna be ok I would like to add some doc too

sure, welcome adding doc. Maybe in separate PR?

**Pull request merged**